src/pool/srv_pool.c

```diff
@@ -5635,38 +5651,74 @@ pool_svc_schedule_reconf(struct pool_svc *svc)
 		return;
 	}
 
-	reconf_cancel_and_wait(reconf);
+	D_ASSERT(&svc->ps_reconf_sched == sched || &svc->ps_rfcheck_sched == sched);
+	sched_cancel_and_wait(sched);
 
-	/* Ended by pool_svc_reconf_ult. */
-	reconf_begin(reconf);
+	sched_begin(sched);
 
 	/*
 	 * An extra svc leader reference is not required, because
 	 * pool_svc_step_down_cb waits for this ULT to terminate.
 	 *
-	 * ULT tracking is achieved through svc->ps_reconf, not a ULT handle.
+	 * ULT tracking is achieved through sched, not a ULT handle.
 	 */
-	rc = dss_ult_create(pool_svc_reconf_ult, svc, DSS_XS_SELF, 0, 0, NULL /* ult */);
+	rc = dss_ult_create(func, svc, DSS_XS_SELF, 0, 0, NULL /* ult */);
 	if (rc != 0) {
-		D_ERROR(DF_UUID": failed to create reconfiguration ULT: "DF_RC"\n",
+		D_ERROR(DF_UUID": failed to create ULT: "DF_RC"\n",
 			DP_UUID(svc->ps_uuid), DP_RC(rc));
-		reconf_end(reconf);
+		sched_end(sched);
 	}
 
 	D_DEBUG(DB_MD, DF_UUID": end: "DF_RC"\n", DP_UUID(svc->ps_uuid), DP_RC(rc));
 }
 
-static void
-pool_svc_cancel_and_wait_reconf(struct pool_svc *svc)
-{
-	reconf_cancel_and_wait(&svc->ps_reconf);
-}
-
 static int pool_find_all_targets_by_addr(struct pool_map *map,
 					 struct pool_target_addr_list *list,
 					 struct pool_target_id_list *tgt_list,
 					 struct pool_target_addr_list *inval);
 
+static int
+cont_rf_check_cb(uuid_t pool_uuid, uuid_t cont_uuid, struct rdb_tx *tx, void *arg)
+{
+	struct pool_svc_sched *sched = arg;
+	int rc;
+
+	/* If anything happened during rf check, let's continue to check the next container
+	 * for the moment.
+	 */
+	rc = ds_cont_rf_check(pool_uuid, cont_uuid, tx);
+	if (rc)
+		D_CDEBUG(rc == -DER_RF, DB_MD, DLOG_ERR, DF_CONT" check_rf: "DF_RC"\n",
+			 DP_CONT(pool_uuid, cont_uuid), DP_RC(rc));
+
+	if (sched->psc_canceled) {
+		D_DEBUG(DB_MD, DF_CONT" is canceled.\n", DP_CONT(pool_uuid, cont_uuid));
+		return 1;
+	}
+
+	return 0;
+}
+
+static void
+pool_svc_rfcheck_ult(void *arg)
+{
+	struct pool_svc	*svc = arg;
+	int rc;
+
+	do {
+		/* retry until some one stop the pool svc(rc == 1) or succeed */
+		rc = ds_cont_rdb_iterate(svc->ps_uuid, cont_rf_check_cb,
+					 &svc->ps_rfcheck_sched);
+		if (rc >= 0)
+			break;
+		D_ERROR(DF_UUID" check rf with %d and retry\n", DP_UUID(svc->ps_uuid), rc);
```

Comment:
As the code is written, cont_rf_check_cb() will always return 1 or 0 and any errors ignored. So this D_ERROR and dss_sleep() call are dead code and won't be reached.

Not sure why a loop is needed, even if the code were reachable though? if some error were to be returned from ds_cont_rdb_iterate and the error condition persisted, then this code would just loop forever (or until this PS replica stepped down)?

Reply:
Yes, the callback only return 1 and 0, but ds_cont_rdb_iterate() might still return < 0, so the loop is to make sure the rf_check will never be missed due to temporary error. 

Hmm, if the PS is stepping  down, then sched will make sure callback return 1 and break?

Reply:
Yeah I see about any ds_cont_rdb_iterate() rc < 0 return values.

Wondering about the chances of an error that is not temporary, repeats rapidly and the D_ERROR in this loop causes log spam.  Would it make sense to sleep longer between tries?

Reply:
Hmm, I am not sure how long should we wait to retry.  Though I think we want to recheck/sync the container property asap if there any failure happened. dss_sleep(0) will sure to yield itself, and make sure other ULT being scheduled.

I can change it to dss_sleep(2) if the patch needs to refresh.